### PR TITLE
Make internal ClusterInfo::getPlan() wait for initial plan load from agency

### PR DESCRIPTION
**Pull request by jsteemann**
### Scope & Purpose

Otherwise the maintenance, fired off by the heartbeat thread, may not wait for the initial data to arrive from the agency, which can cause Plan and Current to be empty when maintenance starts.
No backport for 3.6 is needed, as 3.6 has blocking agency communication here in case the Plan is not there.

- [x] :hankey: Bugfix (requires CHANGELOG entry)
- [ ] :pizza: New feature (requires CHANGELOG entry, feature documentation and release notes)
- [ ] :fire: Performance improvement
- [ ] :hammer: Refactoring/simplification
- [x] :book: CHANGELOG entry made

#### Backports:

- [x] No backports required

### Testing & Verification

- [x] The behavior in this PR was *manually tested*
- [x] This change is already covered by existing tests, such as *mini chaos*.

Link to Jenkins PR run:
http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/13365/

### CODE REVIEW

**Comment by jsteemann** [comment]
http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/13369/

**Comment by jsteemann** [comment]
http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/13373/

**Comment by KVS85** [comment]
Tests blue.

File: arangod/Cluster/ClusterInfo.cpp [hunk 1]
```diff
@@ -5021,7 +5021,7 @@ ClusterInfo::getPlan(uint64_t& index, std::unordered_set<std::string> const& dir
 
   // We should never proceed here, until we have seen an
   // initial agency cache through loadPlan
-  waitForPlan(1);
```
Comment (neunhoef) [inline arangod/Cluster/ClusterInfo.cpp:5024]: We also need to address `waitForCurrent` calls, I found another one with forgotten `.wait()` or `.get()`, it is in line 5054.

**Review by neunhoef — CHANGES_REQUESTED** [review]

**Review by neunhoef — COMMENTED** [review]
Just for the record, here is how it is supposed to work at server start:
  1. AgencyCache and its thread starts polling from the agency. Since it does not have a State yet, it is guaranteed to get the latest snapshot and all that follows it.
  2. The PlanSyncer and CurrentSyncer in `ClusterInfo` request a local callback from AgencyCache whenever /arango/Plan/Version or /arango/Current/Version respectively change, this triggeres `loadPlan` and `loadCurrent` eventually.
  3. `loadPlan` and `loadCurrent` are the only ones which change `_planIndex`, `_planVersion`, `_currentVersion`, `_currentIndex` in `ClusterInfo`.
  4. `getPlan` and `getCurrent` are protected with a `waitForPlan(1).get()` and `waitForCurrent(1).get()` respectively. So nobody outside can get a plan or a current from `ClusterInfo` before it is set up correctly.
  5. Additionally, the `ClusterFeature::start` method has a `waitFor(1).get()` call such that this does not proceed unless the AgencyCache has got a first delivery from the agency, which will already be reasonably up to date. This is all before it even starts the HeartbeatThread (running Maintenance) and the Syncers.
  6. The SyncerThreads do a `loadPlan` or `loadCurrent` respectively before they begin their main loop.

**Review by neunhoef — APPROVED** [review]
LGTM. Further to my previous comment: It is safe to throw in `getPlan` and `getCurrent` when we are shutting down. This is caught in the `run` method of the `HeartbeatBackgrounThread` (which does maintenance via `DBServerAgencySync`) and the shutdown will then work properly. Furthermore, we will not do a maintenance run in this case, even if this is the very first run.

**Comment by jsteemann** [comment]
http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/13381/

**Comment by KVS85** [comment]
`IResearch*` unit tests are broken in the last run.

**Comment by jsteemann** [comment]
http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/13394/

**Comment by KVS85** [comment]
Tests blue.

**Review by KVS85 — APPROVED** [review]
LGTM.

**Review by fceller — APPROVED** [review]
LGTM

**Pull request merged**